dev/release.sh

```diff
@@ -0,0 +1,614 @@
+
+echo "== Initializing work tree"
+
+# Generate a release branch name.  We also use this as part of the cloned
+# directory name
+new_branch=$(eval $(git cat-file blob HEAD:VERSION)
+             echo "openssl-$MAJOR.$MINOR.x")
+release_clone="$new_branch-release-tmp"
+
+if $do_branch && [ "$new_branch" = "$current_branch" ]; then
+    echo >&2 "Warning: --branch ignored, we're already in a release branch"
+    do_branch=false
+fi
+
+echo "== Work tree will be in $release_clone"
+
+# Make a clone in a subdirectory and move there
+if ! [ -d "$release_clone" ]; then
+    $VERBOSE "== Cloning to $release_clone"
+    git clone $git_quiet -b "$current_branch" . "$release_clone"
+fi
+cd "$release_clone"
+
+# Check that we're still on the same branch
+if [ "$current_branch" != "$(git rev-parse --abbrev-ref HEAD)" ]; then
+   echo >&2 "The cloned sub-directory '$release_clone' is on a branch"
+   echo >&2 "other than '$current_branch'"
+   echo >&2 "Please 'cd \"$(pwd)\"; git checkout $current_branch'"
+   exit 1
+fi
+
+SOURCEDIR=$(pwd)
+$DEBUG >&2 "DEBUG: Source directory is $SOURCEDIR"
+
+# Release ############################################################
+
+get_version
+# We always expect to start from a state of development
+if [ "$TYPE" != 'dev' ]; then
+    echo >&2 "Not in a development branch"
+    echo >&2 "Have a look at the git log in $release_clone, it may be that"
+    echo >&2 "a previous crash left it in an intermediate state and that"
+    echo >&2 "need to drop the top commit:"
+    echo >&2 ""
+    echo >&2 "(cd $release_clone; git reset --hard HEAD^)"
+    echo >&2 "# WARNING! LOOK BEFORE YOU ACT"
+    exit 1
+fi
+
+# We only create a release branch if the patch number is zero
+if [ $PATCH -ne 0 ]; then
+    if $do_branch; then
+        echo >&2 "Warning! We're already in a release branch; --branch ignored"
+    fi
+    do_branch=false
+fi
+
+# Update the version information.  This won't save anything anywhere, yet,
+# but does check for possible next_method errors before we do bigger work.
+next_release_state "$next_method"
+
+if $do_branch; then
+    $VERBOSE "== Creating a release branch: $new_branch"
+    git checkout $git_quiet -b "$new_branch"
+fi
+
+echo "== Configuring OpenSSL for update and release.  This make take a bit of time"
+
+./Configure cc | while read L; do $VERBOSE "> $L"; done
+
+$VERBOSE "== Checking source file updates"
+
+make update | while read L; do $VERBOSE "> $L"; done
+
+if [ -n "$(git status --porcelain)" ]; then
+    $VERBOSE "== Committing updates"
+    git add -u
+    git commit $git_quiet -m 'make update'
+fi
+
+# Write the version information we updated
+set_version
+
+if [ -n "$PRE_RELEASE_TAG" ]; then
+    release=$MAJOR.$MINOR.$PATCH-$PRE_RELEASE_TAG$BUILD_METADATA
+    announce_template=openssl-announce-pre-release.tmpl
+else
+    release=$MAJOR.$MINOR.$PATCH$BUILD_METADATA
+    announce_template=openssl-announce-release.tmpl
+fi
+tag="openssl-$release"
```

Comment:
Why the change in tag format? Historically we've used a format like "OpenSSL_1_1_1c"

Reply:
The old tag format is derived from what CVS allowed back in the days.  CVS branches and tags were quite limited, [they only allow letters, digits, dashes and underscores](http://web.mit.edu/gnu/doc/html/cvs_8.html).  Git branches and tags, on the other hand, allow periods, and actually most characters (only a few, like `:`, are disallowed, for syntactic reasons).  So there's really no reason to continue with the period-to-underscore conversion.

Also, I see a benefit in having the same name as the tar file.  Copy-n-paste, just add `.tar.gz` and voilà!

Reply:
I wrote a bit more about it here: https://mta.openssl.org/pipermail/openssl-project/2020-April/001931.html